koala/ai.ts

```diff
@@ -0,0 +1,165 @@
+import {
+  experimental_generateImage as generateImage,
+  generateObject,
+  generateText,
+  experimental_transcribe as transcribe,
+} from "ai";
+import { z } from "zod";
+import { errorReport } from "./error-report";
+import {
+  getDefaultImageModel,
+  getDefaultTextModel,
+  getDefaultTranscriptionModel,
+  registry,
+} from "./provider-registry";
+
+type TextModel = "fast" | "grammar" | "reasoning" | "default";
+type ImageModel = "default" | "fast";
+type LanguageModelIdentifier =
+  | `openai:${TextModel}`
+  | `anthropic:${TextModel}`;
+type ImageModelIdentifier = `openai:${ImageModel}`;
+
+export async function generateAIText(options: {
+  model?: LanguageModelIdentifier;
+  messages: CoreMessage[];
+  temperature?: number;
+  maxTokens?: number;
+}) {
+  const modelId = options.model?.includes(":")
+    ? options.model
+    : options.model
+      ? `openai:${options.model}`
+      : undefined;
+
+  const { text } = await generateText({
+    model: modelId
+      ? registry.languageModel(modelId as LanguageModelIdentifier)
+      : getDefaultTextModel(),
+    messages: options.messages,
+    temperature: options.temperature,
+    maxTokens: options.maxTokens,
+  });
+
+  return text;
+}
+
+export async function generateStructuredOutput<T>(options: {
+  model?: string;
+  messages: CoreMessage[];
+  schema: z.ZodSchema<T>;
+  temperature?: number;
+  maxTokens?: number;
+}) {
+  const other = options.model ? `openai:${options.model}` : undefined;
+  const modelId = options.model?.includes(":") ? options.model : other;
+
+  const { object } = await generateObject({
+    model: modelId
+      ? registry.languageModel(modelId as LanguageModelIdentifier)
+      : getDefaultTextModel(),
+    messages: options.messages,
+    schema: options.schema,
+    temperature: options.temperature,
+    maxTokens: options.maxTokens,
+  });
+
+  return object;
+}
+
+export async function generateAIImage(
+  prompt: string,
+  size?: "1024x1024" | "1792x1024" | "1024x1792",
+  model?: string,
+) {
+  const other = model ? `openai:${model}` : undefined;
+  const modelId = model?.includes(":") ? model : other;
+
+  const { image } = await generateImage({
+    model: modelId
+      ? registry.imageModel(modelId as ImageModelIdentifier)
+      : getDefaultImageModel(),
+    prompt,
+    size: size || "1024x1024",
+  });
+
+  return (
+    (image as { url?: string; data?: string; base64?: string }).url ||
+    (image as { url?: string; data?: string; base64?: string }).data ||
+    (image as { url?: string; data?: string; base64?: string }).base64 ||
+    (typeof image === "string" ? image : "")
+  );
+}
+
+export async function transcribeAudio(
+  audioFile: Buffer | ArrayBuffer,
+  options?: {
+    model?: string;
+    prompt?: string;
+  },
+) {
+  const prompt = options?.prompt || "";
+  const { text } = await transcribe({
+    model: getDefaultTranscriptionModel(options?.model),
+    audio: audioFile,
+    providerOptions: { openai: { prompt } },
```

Comment:
One thing you will want to be aware of when you implement the Bedrock/Anthropic version of this API: it seems to hallucinate the text language about 1-5% of the time without a prompt.

I added the prompt back for GPT-4O-Transcribe, but it seems the option is only available as a vendor-specific override currently. There is no universal prompt option in `experimental_transcribe` right now I don't think.